0000-event-system.md

```diff
@@ -0,0 +1,142 @@
+* Start Date: 2018-01-16
+* RFC PR: (leave this empty)
+* React Navigation Issue: (leave this empty)
+
+# Summary
+
+Add an event emitter to the navigation prop that emits focus/blur events for the route.
+
+This proposal has been pioneered by @satya164: https://github.com/react-navigation/react-navigation/issues/1363 https://github.com/react-navigation/react-navigation/issues/51
+
+# Basic example
+
+In this example, we focus the first text input when the screen has completed transitioning:
+
+```
+componentDidMount() {
+  this._sub = this.props.navigation.addEventListener(
+    'onDidFocus',
+    this._focusFirstTextInput
+  );
+}
+componentWillUnmount() {
+  this._sub.remove();
+}
+```
+
+# Motivation
+
+Screens need to know about navigation changes while they are mounted. Consider the following use cases:
+
+### Data and Media subscription
+
+Screens must know when to request live data or when to enable/disable resource-intensive APIs. Media should be able to auto-pause at correct times.
+
+### Unmounting
+
+Tabs and cards may need to unmount and re-mount contents when they are focused.
+
+May be useful in conjunction with removeClippedSubviews and utilities like https://github.com/SoftwareMansion/react-native-resource-saving-container
+
+### Back button handling
+
+Screens may want to prevent back actions based on internal state
+
+### Keyboard dismissal
+
+KB must dismiss when scene _starts_ to transition back (mayBlur)
+
+### StatusBar Re-configuration
+
+Screens may want to re-configure the status bar when the become active or blurred
+
+### Input autofocus
+
+Inputs want to open the keyboard after the screen finishes transition
+
+### “Refocus”
+
+On extra tab press, navigate to top of stack, then scroll to top on next press
+
+# Detailed design
+
+## Events
+
+An event emitter will be provided as part of each screen navigation prop. The events will be emitted by the navigator as changes happen to the navigation state.
+
+`this.props.navigation.addEventListener` will be a `NavigationEventSubscriber`:
+
+```
+export type NavigationEventSubscriber = (
+  eventName: string,
+  eventHandler: () => void
+) => {
+  +remove: () => void,
+};
+```
+
+The following event will be emitted for _every_ navigation prop:
+
+### onAction(action, newState)
+
+This fires for every action that may affect the navigation state.
```

Comment:
if i'm listening for an action in the profile route within a stack in the home tab, and some navigation action is dispatched from the settings tab, would this fire?

Reply:
nope and I clarified this